File: dynamice programming/2328.py
# Microsoft, Abode
# DFS + Memoization
# Tc = O(m*n)
# Sc = O(m*n)

class Solution:
    def countPaths(self, grid):
        m, n = len(grid), len(grid[0])
        MOD = 10**9 + 7
        directions = [(-1, 0), (0, -1), (0, 1), (1, 0)]
        t = [[-1] * n for _ in range(m)]

        def isSafe(i, j):
            return 0 <= i < m and 0 <= j < n

        def dfs(i, j):
            if t[i][j] != -1:
                return t[i][j]
            
            answer = 1  # Count the current cell
            for dir in directions:
                i_, j_ = i + dir[0], j + dir[1]
                if isSafe(i_, j_) and grid[i_][j_] < grid[i][j]:
                    answer = (answer + dfs(i_, j_)) % MOD
            
            t[i][j] = answer
            return answer

        result = 0
        for i in range(m):
            for j in range(n):
                result = (result + dfs(i, j)) % MOD
        
        return result

File: dynamice programming/2328.js


function countPaths(grid){
    const m = grid.length, n = grid[0].length;
    const MOD = 10**9 + 7;
    const directions = [[-1, 0], [0, -1], [0, 1], [1, 0]];
    const t = Array.from({ length: m }, () => Array(n).fill(-1));

    function isSafe(i, j) {
        return 0 <= i && i < m && 0 <= j && j < n;
    }

    function dfs(i, j) {
        if (t[i][j] !== -1) {
            return t[i][j];
        }
        
        let answer = 1;  // Count the current cell
        for (const dir of directions) {
            const i_ = i + dir[0], j_ = j + dir[1];
            if (isSafe(i_, j_) && grid[i_][j_] < grid[i][j]) {
                answer = (answer + dfs(i_, j_)) % MOD;
            }
        }
        
        t[i][j] = answer;
        return answer;
    }

    let result = 0;
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            result = (result + dfs(i, j)) % MOD;
        }
    }
    
    return result;
}